File: android/aisupport-sdk/src/main/java/io/polydev/aisupport/SocketClient.kt
package io.polydev.aisupport

import io.socket.client.IO
import io.socket.client.Socket
import kotlinx.serialization.json.Json
import org.json.JSONObject
import java.net.URI

/**
 * AISUP Socket Client for Android
 * Handles real-time WebSocket communication with AISup backend
 */
class AISUPSocketClient(private val config: AISUPConfig) {
    
    private var socket: Socket? = null
    private var chatId: String? = null
    
    private val json = Json { 
        ignoreUnknownKeys = true 
        isLenient = true
    }
    
    var status: ConnectionStatus = ConnectionStatus.DISCONNECTED
        private set
    
    // Callbacks
    var onConnect: (() -> Unit)? = null
    var onDisconnect: ((String) -> Unit)? = null
    var onError: ((Throwable) -> Unit)? = null
    var onMessage: ((Message) -> Unit)? = null
    var onChatUpdated: ((Chat) -> Unit)? = null
    var onTyping: ((Boolean) -> Unit)? = null
    var onStatusChange: ((ConnectionStatus) -> Unit)? = null
    
    val isConnected: Boolean
        get() = status == ConnectionStatus.CONNECTED && socket?.connected() == true
    
    /**
     * Connect to WebSocket server
     */
    fun connect() {
        if (socket?.connected() == true) return
        
        setStatus(ConnectionStatus.CONNECTING)
        
        try {
            val options = IO.Options().apply {
                forceNew = true
                reconnection = true
                reconnectionAttempts = 5
                reconnectionDelay = 1000
                auth = mapOf("apiKey" to config.apiKey)
                extraHeaders = mapOf("X-API-Key" to listOf(config.apiKey))
            }
            
            socket = IO.socket(URI.create(config.wsUrl), options)
            setupEventListeners()
            socket?.connect()
        } catch (e: Exception) {
            setStatus(ConnectionStatus.ERROR)
            onError?.invoke(e)
        }
    }
    
    /**
     * Disconnect from WebSocket server
     */
    fun disconnect() {
        socket?.disconnect()
        socket = null
        setStatus(ConnectionStatus.DISCONNECTED)
    }
    
    /**
     * Join a chat room to receive messages
     */
    fun joinChat(chatId: String, callback: (Result<String>) -> Unit) {
        if (!isConnected) {
            callback(Result.failure(AISUPException.SocketError("Socket not connected")))
            return
        }
        
        this.chatId = chatId
        
        socket?.emit("integration_join", JSONObject().apply {
            put("chatId", chatId)
        }) { args ->
            val response = args.getOrNull(0) as? JSONObject
            val status = response?.optString("status")
            
            if (status == "ok") {
                val resultChatId = response?.optString("chatId") ?: chatId
                callback(Result.success(resultChatId))
            } else {
                val message = response?.optString("message") ?: "Failed to join chat"
                callback(Result.failure(AISUPException.SocketError(message)))
            }
        }
    }
    
    /**
     * Leave current chat room
     */
    fun leaveChat(callback: (() -> Unit)? = null) {
        val currentChatId = chatId ?: run {
            callback?.invoke()
            return
        }
        
        socket?.emit("integration_leave", JSONObject().apply {
            put("chatId", currentChatId)
        }) { _ ->
            this.chatId = null
            callback?.invoke()
        }
    }
    
    /**
     * Send typing indicator
     */
    fun sendTyping(isTyping: Boolean) {
        val currentChatId = chatId ?: return
        if (!isConnected) return
        
        socket?.emit("typing", JSONObject().apply {
            put("chatId", currentChatId)
            put("isTyping", isTyping)
        })
    }
    
    private fun setStatus(newStatus: ConnectionStatus) {
        status = newStatus
        onStatusChange?.invoke(newStatus)
    }
    
    private fun setupEventListeners() {
        socket?.on(Socket.EVENT_CONNECT) {
            setStatus(ConnectionStatus.CONNECTED)
            onConnect?.invoke()
            
            // Rejoin chat if we were in one
            chatId?.let { id ->
                joinChat(id) { }
            }
        }
        
        socket?.on(Socket.EVENT_DISCONNECT) { args ->
            setStatus(ConnectionStatus.DISCONNECTED)
            val reason = args.getOrNull(0)?.toString() ?: "Unknown reason"
            onDisconnect?.invoke(reason)
        }
        
        socket?.on(Socket.EVENT_CONNECT_ERROR) { args ->
            setStatus(ConnectionStatus.ERROR)
            val error = args.getOrNull(0) as? Exception 
                ?: Exception("Connection error")
            onError?.invoke(error)
        }
        
        socket?.on("message_added") { args ->
            try {
                val data = args.getOrNull(0) as? JSONObject ?: return@on
                val messageJson = data.getJSONObject("message").toString()
                val message = json.decodeFromString<Message>(messageJson)
                onMessage?.invoke(message)
            } catch (e: Exception) {
                onError?.invoke(e)
            }
        }
        
        socket?.on("chat_updated") { args ->
            try {
                val data = args.getOrNull(0) as? JSONObject ?: return@on
                val chatJson = data.getJSONObject("chat").toString()
                val chat = json.decodeFromString<Chat>(chatJson)
                onChatUpdated?.invoke(chat)
            } catch (e: Exception) {
                onError?.invoke(e)
            }
        }
        
        socket?.on("typing") { args ->
            try {
                val data = args.getOrNull(0) as? JSONObject ?: return@on
                val isTyping = data.getBoolean("isTyping")
                onTyping?.invoke(isTyping)
            } catch (e: Exception) {
                // Ignore typing errors
            }
        }
    }
}

File: src/socket-client.ts
/**
 * AISup Socket Client
 * Standalone WebSocket client for real-time communication with AISup backend
 * Can be used without the UI widget
 */

import { io, Socket } from 'socket.io-client';
import type { AISupportConfig, Message, Chat } from './api-client';

export type ConnectionStatus = 'disconnected' | 'connecting' | 'connected' | 'error';

export interface SocketEventHandlers {
  onConnect?: () => void;
  onDisconnect?: (reason: string) => void;
  onError?: (error: Error) => void;
  onMessage?: (message: Message) => void;
  onChatUpdated?: (chat: Chat) => void;
  onTyping?: (isTyping: boolean) => void;
  onConnectionChange?: (status: ConnectionStatus) => void;
}

export class AISupportSocketClient {
  private config: AISupportConfig;
  private socket: Socket | null = null;
  private chatId: string | null = null;
  private handlers: SocketEventHandlers = {};
  private status: ConnectionStatus = 'disconnected';
  private reconnectAttempts = 0;
  private maxReconnectAttempts = 5;

  constructor(config: AISupportConfig) {
    this.config = {
      ...config,
      wsUrl: config.wsUrl || config.apiUrl,
    };
  }

  /**
   * Get current connection status
   */
  getStatus(): ConnectionStatus {
    return this.status;
  }

  /**
   * Check if socket is connected
   */
  isConnected(): boolean {
    return this.status === 'connected' && this.socket?.connected === true;
  }

  /**
   * Set event handlers
   */
  setHandlers(handlers: SocketEventHandlers): void {
    this.handlers = { ...this.handlers, ...handlers };
  }

  /**
   * Connect to WebSocket server
   */
  connect(): void {
    if (this.socket?.connected) {
      return;
    }

    this.setStatus('connecting');

    this.socket = io(this.config.wsUrl!, {
      auth: { apiKey: this.config.apiKey },
      transports: ['websocket', 'polling'],
      reconnection: true,
      reconnectionAttempts: this.maxReconnectAttempts,
      reconnectionDelay: 1000,
      reconnectionDelayMax: 5000,
    });

    this.setupEventListeners();
  }

  /**
   * Disconnect from WebSocket server
   */
  disconnect(): void {
    if (this.socket) {
      this.socket.disconnect();
      this.socket = null;
    }
    this.setStatus('disconnected');
  }

  /**
   * Join a chat room to receive messages
   */
  async joinChat(chatId: string): Promise<{ status: string; chatId?: string; message?: string }> {
    if (!this.socket?.connected) {
      throw new Error('Socket not connected. Call connect() first.');
    }

    this.chatId = chatId;

    return new Promise((resolve, reject) => {
      this.socket!.emit('integration_join', { chatId }, (response: { status: string; chatId?: string; message?: string }) => {
        if (response.status === 'ok') {
          resolve(response);
        } else {
          reject(new Error(response.message || 'Failed to join chat'));
        }
      });
    });
  }

  /**
   * Leave current chat room
   */
  async leaveChat(): Promise<void> {
    if (!this.socket?.connected || !this.chatId) {
      return;
    }

    return new Promise((resolve) => {
      this.socket!.emit('integration_leave', { chatId: this.chatId }, () => {
        this.chatId = null;
        resolve();
      });
    });
  }

  /**
   * Send typing indicator
   */
  sendTyping(isTyping: boolean): void {
    if (!this.socket?.connected || !this.chatId) {
      return;
    }

    this.socket.emit('typing', { chatId: this.chatId, isTyping });
  }

  private setStatus(status: ConnectionStatus): void {
    this.status = status;
    this.handlers.onConnectionChange?.(status);
  }

  private setupEventListeners(): void {
    if (!this.socket) return;

    this.socket.on('connect', () => {
      this.reconnectAttempts = 0;
      this.setStatus('connected');
      this.handlers.onConnect?.();

      // Rejoin chat if we were in one
      if (this.chatId) {
        this.joinChat(this.chatId).catch(console.error);
      }
    });

    this.socket.on('disconnect', (reason) => {
      this.setStatus('disconnected');
      this.handlers.onDisconnect?.(reason);
    });

    this.socket.on('connect_error', (error) => {
      this.reconnectAttempts++;
      if (this.reconnectAttempts >= this.maxReconnectAttempts) {
        this.setStatus('error');
      }
      this.handlers.onError?.(error);
    });

    this.socket.on('message_added', ({ message }: { chatId: string; message: Message }) => {
      this.handlers.onMessage?.(message);
    });

    this.socket.on('chat_updated', ({ chat }: { chatId: string; chat: Chat }) => {
      this.handlers.onChatUpdated?.(chat);
    });

    this.socket.on('typing', ({ isTyping }: { isTyping: boolean }) => {
      this.handlers.onTyping?.(isTyping);
    });
  }
}

export default AISupportSocketClient;
